server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -244,7 +244,7 @@ public enum MetadataVersion {
      * <strong>Think carefully before you update this value. ONCE A METADATA VERSION IS PRODUCTION,
      * IT CANNOT BE CHANGED.</strong>
      */
-    public static final MetadataVersion LATEST_PRODUCTION = IBP_3_8_IV0;
+    public static final MetadataVersion LATEST_PRODUCTION = IBP_3_9_IV0;
```

Comment:
Hmm, ListOffsetRequest has latestVersionUnstable set to true. So, it's kind of weird to have a production MV depending on a unstable PRC.

@clolov @showuon @chia7712 : Is KIP-1005 done? Could we set latestVersionUnstable to false for ListOffsetRequest in 3.9?

Reply:
> Is KIP-1005 done? Could we set latestVersionUnstable to false for ListOffsetRequest in 3.9?

KIP-1005 is completed and backport to 3.9, so yes latestVersionUnstable should be false now.

Reply:
@junrao : yes, KIP-1005 will ship in AK 3.9.

Reply:
@chia7712 do you plan to update https://github.com/apache/kafka/blob/df04887ba597941674a28d54bd35202bbd21631e/clients/src/main/resources/common/message/ListOffsetsRequest.json#L42

Reply:
> do you plan to update 

I assumed @cmccabe will update it in this PR :smile: 

Will file a jira and PR for it later

Reply:
> It's a bit surprising that no test failed with a production MV depending on an unstable ListOffset RPC. Do you know why?

I don't think there is the "connection" between MV and unstable ListOffset RPC. The only condition used to enable unstable RPC is to set `unstable.api.versions.enable=true` regardless of MV.

For another, the version supports of ListOffset is a bit weird. We don't add the version check on both client/server side for the "specific" timestamp. 

For example: version 8 (KIP-405) added `EARLIEST_LOCAL_TIMESTAMP=-4` flag. the question are: 

1. Should server return `UNSUPPORTED_VERSION` if server receive a request with `version<8` and `timestamp=-4`?
2. Should admin check the timestamp and throw unsupported version if the server is unsupported to that timestamp flag?

It seems to me the version check is necessary to make 1) newer server can return suitable error to older client and 2) newer client can generate suitable error itself rather than receive undefined result from older server. 

@junrao WDYT?

Reply:
I agree we should add tests for it.

Reply:
@chia7712 : Thanks for the reply. Previously, Christo's PR seems to have caused some tests to fail because of this (https://github.com/apache/kafka/pull/15673#issuecomment-2104306940). So, I was wondering why the failure didn't show up here.

Reply:
> Christo's PR seems to have caused some tests to fail because of this (https://github.com/apache/kafka/pull/15673#issuecomment-2104306940). So, I was wondering why the failure didn't show up here.

the CI is gone so I can't see "which test" you discussed. According to https://github.com/apache/kafka/commit/53276b12c57f325fc62e0133eae1eae023c3b74b,  I guess the test case is `testApiVersionsRequestIncludesUnreleasedApis`. It is used to verify the api key and version range. Everything is good but #15761 introduced a "wrong" config to it. However, the test was not failed since we did not have "unstable" version yet.

That is why #15673 changed the `unstable.api.versions.enable` from `false` to `true`, because it added a unstable "version=0" to LIST_OFFSET.

In short, I feel the test case `testApiVersionsRequestIncludesUnreleasedApis` is NOT related to MV. If we want to build the relationship between "unstable version" and "MV". Maybe we can add new field to `ApiKeys` to declare the MV for the RPC having "unstable version". And then we can add tests for following checks

1. the RPC having unstable version must declare a MV in `ApiKeys`. The MV is equal to the release which can change the version from unstable to "stable"
2. the `MetadataVersion#LATEST_PRODUCTION` must be smaller than MV in `ApiKeys`. That ensures `LATEST_PRODUCTION` will NOT depend on unstable versions.

This idea does NOT impact the production code, but it increase the complexity of updating MV/version.

@junrao WDYT?

Reply:
> For example: version 8 (KIP-405) added EARLIEST_LOCAL_TIMESTAMP=-4 flag. the question are:
    Should server return UNSUPPORTED_VERSION if server receive a request with version<8 and timestamp=-4?
    Should admin check the timestamp and throw unsupported version if the server is unsupported to that timestamp flag?
It seems to me the version check is necessary to make 1) newer server can return suitable error to older client and 2) newer client can generate suitable error itself rather than receive undefined result from older server.

I open https://issues.apache.org/jira/browse/KAFKA-17331 for it. @cmccabe @junrao Sorry that I feel this is a part of KIP-1005. we will complete it ASAP

Reply:
I don't think MV needs to tie to unstable versions. We used to not have unstable metadata versions in a release, but now we allow it if the MV is for the next release and we have the mechanisms to do this. 

I think generally, the unstable api version in a release is OK, BUT if we say the KIP is completed in the release, the api should be stable (unless we doing using early access etc). I don't think we need to tie api versions to MV at this time.

Reply:
>  I don't think we need to tie api versions to MV at this time.

@jolshan : Just to be clear. I think the issue is that some inter broker requests (e.g. replication) depend on MV to set the request version, not based on supported versions returned in ApiVersion. If we have a stable MV depending on an unstable version of an inter-broker RPC, in the default setting (with stable MV and requests), the broker will receive a newer version of the request that it doesn't support.

@chia7712 : For this particular case, `ListOffset` is an inter broker RPC. In the default setting, the replication code will use v9 of `ListOffset`. However, by default, v9 of `ListOffset` is not supported with latestVersionUnstable set to true. So, I still don't understand why some existing tests don't fail. We probably need to understand the testing gap. If we have sufficient test coverage, we probably don't need additional protocols.

Reply:
Yes -- sorry I was imprecise. If the api version is advertised by a MV that is stable it sould also be stable.

Reply:
> However, by default, v9 of ListOffset is not supported with latestVersionUnstable set to true. So, I still don't understand why some existing tests don't fail. We probably need to understand the testing gap. If we have sufficient test coverage, we probably don't need additional protocols.

yes, you are right. we should have a test for replica path. I just write a IT which can produce following error when LIST_OFFSET has unstable version:
```
[2024-08-14 03:27:13,119] ERROR Closing socket for 127.0.0.1:36885-127.0.0.1:48854-3 because of error (kafka.network.Processor:76)
org.apache.kafka.common.errors.InvalidRequestException: Received request api key LIST_OFFSETS with version 9 which is not enabled
```
Also, it seems we create IT with unstable APIs always [0] and I guess that is why we don't see failed tests.

**Summary**
1. broker-to-broker: the LIST_OFFSET version is based on MV [1], and we use IT to protect us from updating MV with an unstable version (https://issues.apache.org/jira/browse/KAFKA-17336)
2. new-client-to-old-broker: the builder of LIST_OFFSET will check the timestamp and its matched version (https://issues.apache.org/jira/browse/KAFKA-17331)
3. old-client-to-new-broker: the broker will check the timestamp and its matched version (https://issues.apache.org/jira/browse/KAFKA-17331)


> I didn't realize ListOffset was being used in the replication path. I guess this is just for tiered storage.

@jolshan not really, the request is used to truncate in the replication path. I have attached a sample code in https://issues.apache.org/jira/browse/KAFKA-17336. we will complete it tomorrow.

[0] https://github.com/apache/kafka/blob/8d29bc1fa8ce75b6807b957ea35ad39e8dd9e04a/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L307
[1] https://github.com/apache/kafka/blob/8d29bc1fa8ce75b6807b957ea35ad39e8dd9e04a/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L474

Reply:
Thanks. https://issues.apache.org/jira/browse/KAFKA-17336 should be good to not release with unstable versions that break api usage.

Reply:
> Also, it seems we create IT with unstable APIs always [0] and I guess that is why we don't see failed tests.

Thanks for the investigation, @chia7712. Maybe we just need to disable unstable API/MV in a few existing integration tests.

Reply:
> Maybe we just need to disable unstable API/MV in a few existing integration tests.

yep, we are looking for the similar tests